### PR TITLE
Factory and corrected parts of the classes

### PR DESCRIPTION
**Pull request by kmuzalewska**
Corrected JPetHit ( checkConsistency ) and add test to non default constructor, JPetBaseSignal ( inNullObject ), JPetPhysSignal ( isNullObject )
Added functions to make objects with all posible parameters. It is done to : JPetScin, JPetFEB, JPetLayer, JPertHit, JPetSigCh, JPetBarrelSlot, JPetTimeWindow, JPetPM, JPetBaseSignal, JPetPhysSignal, JPetRawSignal, JPetTOMBChannel.

### CODE REVIEW

File: JPetParamAndDateFactory/JPetParamAndDateFactory.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+ *  limitations under the License.
+ *
+ *  @file JPetParamAndDateFactory.cpp
+ */
+
+#include "./JPetParamAndDateFactory.h"
+
+namespace factory
+{
+JPetScin makeScin(int id, float attenLen, float length, float height, float width, JPetBarrelSlot& p_barrelSlot )
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: shouldn't it ibe const JPetBarrelSlot& ?

File: JPetParamAndDateFactory/JPetParamAndDateFactory.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+{
+  JPetScin scinObject(id, attenLen, length, height, width);
+  scinObject.setBarrelSlot(p_barrelSlot);
+  return scinObject;
+}
+
+JPetFEB makeFEB(int p_id, bool p_isActive, std::string p_status, std::string p_description, int p_version, int p_userId, int p_n_time_outputs_per_input, int p_n_notime_outputs_per_input, JPetTRB& p_TRB)
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: const std::string& ...

File: JPetParamAndDateFactory/JPetParamAndDateFactory.cpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+JPetLayer makeLayer(int id, bool isActive, std::string name, float radius, JPetFrame& frame)
+{
+  JPetLayer objectLayer(id, isActive, name, radius);
+  objectLayer.setFrame(frame);
+  return objectLayer;
+}
+
+JPetHit makeHit(float e, float qe, float t, float qt, TVector3& pos, JPetPhysSignal& siga, JPetPhysSignal& sigb, JPetBarrelSlot& bslot, JPetScin& scin, float qtd, float td)
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: consts missing

File: JPetParamAndDateFactory/JPetParamAndDateFactory.cpp [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+JPetFEB makeFEB(int p_id, bool p_isActive, std::string p_status, std::string p_description, int p_version, int p_userId, int p_n_time_outputs_per_input, int p_n_notime_outputs_per_input, JPetTRB& p_TRB)
+{
+  JPetFEB objectFEB(p_id, p_isActive, p_status, p_description, p_version, p_userId, p_n_time_outputs_per_input, p_n_notime_outputs_per_input);
+  objectFEB.setTRB(p_TRB);
+  return objectFEB;
+}
+
+JPetLayer makeLayer(int id, bool isActive, std::string name, float radius, JPetFrame& frame)
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: const missing

File: JPetParamAndDateFactory/JPetParamAndDateFactory.cpp [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+   sigChObject.setTRB(trb);
+   sigChObject.setFEB(feb);
+   sigChObject.setTOMBChannel(channel);
+   sigChObject.setThreshold(thr);
+   sigChObject.setDAQch(daqch);
+   sigChObject.setThresholdNumber(threshold_number);
+   return sigChObject;
+  }
+
+JPetBarrelSlot makeBarrelSlot(JPetLayer& p_layer, int id, bool isActive, std::string name, float theta, int inFrameID)
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: const p_payer and std::string

File: JPetParamAndDateFactory/JPetParamAndDateFactory.cpp [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+{
+  JPetBarrelSlot barrelSlotObject(id, isActive, name, theta, inFrameID);
+  barrelSlotObject.setLayer(p_layer);
+  return barrelSlotObject;
+}
+
+JPetTimeWindow makeTimeWindow(JPetSigCh& new_ch, unsigned int index)
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: const

File: JPetParamAndDateFactory/JPetParamAndDateFactory.cpp [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+{
+  JPetTimeWindow timeWindowObject;
+  timeWindowObject.addCh(new_ch);
+  timeWindowObject.setIndex(index);
+  return timeWindowObject;
+}
+
+JPetPM makePM(JPetPM::Side side, int id, int set, int opt, std::pair<float, float>& gain, JPetFEB& p_FEB, JPetScin& p_scin, JPetBarrelSlot& p_barrelSlot)
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: again const missing

File: JPetParamAndDateFactory/JPetParamAndDateFactory.cpp [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+  tombChannelObject.setFEB(p_FEB);
+  tombChannelObject.setTRB(p_TRB);
+  tombChannelObject.setPM(p_PM);
+  tombChannelObject.setThreshold(p_threshold);
+  tombChannelObject.setLocalChannelNumber(lcn);
+  tombChannelObject.setFEBInputNumber(fin);
+  return tombChannelObject;
+}
+
+}
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: newline

File: JPetParamAndDateFactory/JPetParamAndDateFactory.h [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ *  you may not use this file except in compliance with the License.
+ *  You may find a copy of the License in the LICENCE file.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  @file JPetParamAndDateFactory.h
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.h]: It should be called ParamAndDataFactory not  Date,
Date - 1th of May
Data - well data :)

File: JPetParamAndDateFactory/JPetParamAndDateFactory.h [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+#include "../JPetPM/JPetPM.h"
+#include "../JPetTRB/JPetTRB.h"
+#include "../JPetFrame/JPetFrame.h"
+#include "../JPetPhysSignal/JPetPhysSignal.h"
+#include "../JPetBaseSignal/JPetBaseSignal.h"
+#include "../JPetRawSignal/JPetRawSignal.h"
+#include  "TVector3.h"
+#include "../JPetTOMBChannel/JPetTOMBChannel.h"
+
+namespace factory
```
Comment (wkrzemien) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.h]: I would change to param_and_data_factory

File: JPetParamAndDateFactory/JPetParamAndDateFactory.cpp (hunk 1 above)
Comment (kmuzalewska) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: i've added the same parameters as it was in base class in set functions, but if it should be another, i'll change

Comment (kmuzalewska) [inline JPetParamAndDateFactory/JPetParamAndDateFactory.cpp]: setBarrelSlot takes argument without const, so it is not compiling, before that modification set function should be change in the class

**Comment by wkrzemien** [comment]
You are right about const, since it is set it cannot be guaranteed to be unmutable. Please ignore const to JPEt class comments, however std::string should be const  ref

File: JPetParamAndDataFactory/JPetParamAndDataFactory.h [hunk 11]
```diff
@@ -32,7 +32,7 @@
 #include  "TVector3.h"
 #include "../JPetTOMBChannel/JPetTOMBChannel.h"
 
-namespace factory
+namespace param_and_data_factory
 {
 JPetScin makeScin(int id, float attenLen, float length, float height, float width, JPetBarrelSlot& p_barrelSlot ); 
 JPetFEB makeFEB(int p_id, bool p_isActive, std::string p_status, std::string p_description, int p_version, int p_userId, int p_n_time_outputs_per_input, int p_n_notime_outputs_per_input, JPetTRB& p_TRB);
```
Comment (wkrzemien) [inline JPetParamAndDataFactory/JPetParamAndDataFactory.h]: std::string -> const std::string&  everywhere

File: JPetParamAndDataFactory/JPetParamAndDataFactory.h [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+#include  "TVector3.h"
+#include "../JPetTOMBChannel/JPetTOMBChannel.h"
+
+namespace param_and_data_factory
+{
+JPetScin makeScin(int id, float attenLen, float length, float height, float width, JPetBarrelSlot& p_barrelSlot ); 
+JPetFEB makeFEB(int p_id, bool p_isActive, const std::string& p_status, const std::string& p_description, int p_version, int p_userId, int p_n_time_outputs_per_input, int p_n_notime_outputs_per_input, JPetTRB& p_TRB);
+JPetLayer makeLayer(int id, bool isActive, const std::string& name, float radius,JPetFrame& frame); 
+JPetHit makeHit(float e, float qe, float t, float qt, TVector3& pos, JPetPhysSignal& siga, JPetPhysSignal& sigb, JPetBarrelSlot& bslot, JPetScin& scin, float qtd, float td);
+JPetSigCh makeSigCh(const JPetPM & pm, const JPetTRB & trb, const JPetFEB & feb, const JPetTOMBChannel & channel, float val, JPetSigCh::EdgeType type, float thr, Int_t daqch, unsigned int threshold_number);
```
Comment (wkrzemien) [inline JPetParamAndDataFactory/JPetParamAndDataFactory.h]: You should remove const from objects.

File: JPetParamAndDataFactory/JPetParamAndDataFactory.h [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+JPetFEB makeFEB(int p_id, bool p_isActive, const std::string& p_status, const std::string& p_description, int p_version, int p_userId, int p_n_time_outputs_per_input, int p_n_notime_outputs_per_input, JPetTRB& p_TRB);
+JPetLayer makeLayer(int id, bool isActive, const std::string& name, float radius,JPetFrame& frame); 
+JPetHit makeHit(float e, float qe, float t, float qt, TVector3& pos, JPetPhysSignal& siga, JPetPhysSignal& sigb, JPetBarrelSlot& bslot, JPetScin& scin, float qtd, float td);
+JPetSigCh makeSigCh(const JPetPM & pm, const JPetTRB & trb, const JPetFEB & feb, const JPetTOMBChannel & channel, float val, JPetSigCh::EdgeType type, float thr, Int_t daqch, unsigned int threshold_number); 
+JPetBarrelSlot makeBarrelSlot(JPetLayer& p_layer, int id, bool isActive, const std::string& name, float theta, int inFrameID); 
+JPetTimeWindow makeTimeWindow(JPetSigCh& new_ch, unsigned int index); 
+JPetPM makePM(JPetPM::Side side, int id, int set, int opt, std::pair<float, float>& gain, JPetFEB& p_FEB, JPetScin& p_scin, JPetBarrelSlot& p_barrelSlot); 
+JPetBaseSignal makeBaseSignal(unsigned int index, const JPetPM & pm, const JPetBarrelSlot & bs);
```
Comment (wkrzemien) [inline JPetParamAndDataFactory/JPetParamAndDataFactory.h]: here as well? remove const ?

File: JPetParamAndDataFactory/JPetParamAndDataFactory.h [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+JPetPhysSignal makePhysSignal(float time, float qualityOfTime, double phe, double qualityOfPhe, const JPetRecoSignal& recoSignal);
+JPetRawSignal makeRawSignal(const int points, const JPetSigCh& sigch);
```
Comment (wkrzemien) [inline JPetParamAndDataFactory/JPetParamAndDataFactory.h]: const int points -> int points

File: JPetParamAndDataFactory/JPetParamAndDataFactory.h [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+#include "../JPetTimeWindow/JPetTimeWindow.h"
+#include "../JPetPM/JPetPM.h"
+#include "../JPetTRB/JPetTRB.h"
+#include "../JPetFrame/JPetFrame.h"
+#include "../JPetPhysSignal/JPetPhysSignal.h"
+#include "../JPetBaseSignal/JPetBaseSignal.h"
+#include "../JPetRawSignal/JPetRawSignal.h"
+#include  "TVector3.h"
+#include "../JPetTOMBChannel/JPetTOMBChannel.h"
+
```
Comment (wkrzemien) [inline JPetParamAndDataFactory/JPetParamAndDataFactory.h]: Please add doxygen-style description e.g.
"Set of factory methods to create fully operational parameter and data objects"

Comment (kmuzalewska) [inline JPetParamAndDataFactory/JPetParamAndDataFactory.h]: added

File: JPetParamAndDataFactory/JPetParamAndDataFactory.cpp [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+  JPetPhysSignal physSignalObject;
+  physSignalObject.setTime(time);
+  physSignalObject.setQualityOfTime(qualityOfTime);
+  physSignalObject.setPhe(phe);
+  physSignalObject.setQualityOfPhe(qualityOfPhe);
+  physSignalObject.setRecoSignal(recoSignal);
+  return physSignalObject;
+}
+
+JPetRawSignal makeRawSignal(int points, JPetSigCh& sigch)
```
Comment (wkrzemien) [inline JPetParamAndDataFactory/JPetParamAndDataFactory.cpp]: sorry I havent spotted this earlier. 
We need probably: JPetRawSignal(const std::vector<JPetSigCh>& vec) {
    JPetRawSignal rawSignalObject(vec.size());
for(const auto& sigch: vec) {
  rawSignalObject.addPoint(sigch);
}
  return rawSignalObject;
}

File: JPetParamAndDataFactory/JPetParamAndDataFactory.cpp [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+JPetBarrelSlot makeBarrelSlot(JPetLayer& p_layer, int id, bool isActive, const std::string& name, float theta, int inFrameID)
+{
+  JPetBarrelSlot barrelSlotObject(id, isActive, name, theta, inFrameID);
+  barrelSlotObject.setLayer(p_layer);
+  return barrelSlotObject;
+}
+
+JPetTimeWindow makeTimeWindow(JPetSigCh& new_ch, unsigned int index)
```
Comment (wkrzemien) [inline JPetParamAndDataFactory/JPetParamAndDataFactory.cpp]: the same here, change to 
`makeTimeWindow(const std::vector<JPetSigCh>& vec, unsigned int window_id) {
JPetTimeWindow timeWindowObject;
  for loop  to fill all sigch --> with time
 timeWindowObject.setIndex(window_id);
  return timeWindowObject;
}`

**Pull request merged**